aws-lc-rs/src/cipher/streaming.rs

```diff
@@ -0,0 +1,840 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+use crate::cipher::{
+    Algorithm, DecryptionContext, EncryptionContext, OperatingMode, UnboundCipherKey,
+};
+use crate::error::Unspecified;
+use crate::ptr::{LcPtr, Pointer};
+use aws_lc::{
+    EVP_CIPHER_CTX_new, EVP_CIPHER_iv_length, EVP_CIPHER_key_length, EVP_DecryptFinal_ex,
+    EVP_DecryptInit_ex, EVP_DecryptUpdate, EVP_EncryptFinal_ex, EVP_EncryptInit_ex,
+    EVP_EncryptUpdate, EVP_CIPHER_CTX,
+};
+use std::ptr::null_mut;
+
+/// A key for streaming encryption operations.
+pub struct StreamingEncryptingKey {
+    algorithm: &'static Algorithm,
+    mode: OperatingMode,
+    cipher_ctx: LcPtr<EVP_CIPHER_CTX>,
+    context: EncryptionContext,
+}
+
+/// A struct indicating the portion of a buffer written to, and/or not written to, during an
+/// encryption/decryption operation.
+pub struct BufferUpdate<'a> {
+    written: &'a [u8],
+    remainder: &'a mut [u8],
+}
+
+impl<'a> BufferUpdate<'a> {
+    fn new(out_buffer: &'a mut [u8], written_len: usize) -> Self {
+        let (written, remainder) = out_buffer.split_at_mut(written_len);
+        Self { written, remainder }
+    }
+}
+
+impl BufferUpdate<'_> {
+    /// Returns the slice from the buffer that was modified by the operation.
+    #[must_use]
+    pub fn written(&self) -> &[u8] {
+        self.written
+    }
+
+    /// Returns the slice of the buffer that was not modified by the operation.
+    #[must_use]
+    pub fn remainder(&self) -> &[u8] {
+        self.remainder
+    }
+
+    /// Returns a mutable slice of the buffer that was not modified by the operation.
+    #[must_use]
+    pub fn remainder_mut(&mut self) -> &mut [u8] {
+        self.remainder
+    }
+}
+
+impl StreamingEncryptingKey {
+    #[allow(clippy::needless_pass_by_value)]
+    fn new(
+        key: UnboundCipherKey,
+        mode: OperatingMode,
+        context: EncryptionContext,
+    ) -> Result<Self, Unspecified> {
+        let algorithm = key.algorithm();
+        let cipher_ctx = LcPtr::new(unsafe { EVP_CIPHER_CTX_new() })?;
+        let cipher = mode.evp_cipher(key.algorithm);
+        let key_bytes = key.key_bytes.as_ref();
+        debug_assert_eq!(
+            key_bytes.len(),
+            <usize>::try_from(unsafe { EVP_CIPHER_key_length(*cipher) }).unwrap()
+        );
+        let iv = <&[u8]>::try_from(&context)?;
+        debug_assert_eq!(
+            iv.len(),
+            <usize>::try_from(unsafe { EVP_CIPHER_iv_length(*cipher) }).unwrap()
+        );
+
+        if 1 != unsafe {
+            EVP_EncryptInit_ex(
+                cipher_ctx.as_mut_ptr(),
+                *cipher,
+                null_mut(),
+                key_bytes.as_ptr(),
+                iv.as_ptr(),
```

Comment:
Might be worth leaving a comment behind to indicate that AWS-LC effectively copies the key and initalization vector and keeps it within the EVP context. So the references we pass here are allowed to drop after this call.